File: src/Engine/Core/Pallet.cpp
#include "Pallet.h"
#include "../Invariants/GameInvariants.h"

//protected:
//    //--------------
//    // -- FIELDS --
//    //--------------
//    HashMap<int, int> colors;


int Pallet::color2Int(const int& r, const int& g, const int& b){
    int res;
    res = r; res<<=8;
    res|= g; res<<=8;
    res|= b; // ignore alpha
    //return r<<16 | g<<8 | b;
    return res;
}

int Pallet::color2Int(const sf::Color& color){
    return color2Int(color.r,color.g,color.b);
}

int Pallet::color2Int(IntVector3D color){
    return color2Int(color.x,color.y,color.z);
}

sf::Color Pallet::int2Color(int val){
    sf::Color res;
    res.r = (val>>16);
    res.g = (val>>8)&0xff;
    res.b = val&0xff;
    return res;
}

//--------------------
// -- CONSTRUCTORS --
//--------------------
Pallet::Pallet(const std::string& Filename):AFileResolver(GameInvariants::PALLET_VERSION){
    setFileName(Filename);
}

Pallet::Pallet(DataInputStream& dis):AFileResolver(GameInvariants::PALLET_VERSION){ //throws IOException
    readData(dis);
}
Pallet::Pallet():AFileResolver(GameInvariants::PALLET_VERSION){}


//---------------
// -- METHODS --
//---------------


 void Pallet::applyToImage(const sf::Image& imgIn, sf::Image& imgOut){
    int width = imgIn.GetWidth();
    int height = imgIn.GetHeight();
    imgOut.Create(width, height, sf::Color(0,0,0,0));
    for(int x=0; x<width; ++x){
        for(int y=0; y<height; ++y){
            sf::Color c = imgIn.GetPixel(x,y);
            // ignore transparent pixels
            if (c.a == 0) continue;
            // get the integer representation,
            int original = color2Int(c);
            sf::Color newColor = c;
            if (colors.containsKey(original)){
                // get the matching color
                int matching = colors[original];
                newColor = int2Color(matching);
            }
            newColor.a = c.a;   // keep transparency
            // set the pixel in the destination image
            imgOut.SetPixel(x,y,newColor);
        }
    }
}


//void Pallet::applyToFramesCollection(const FramesCollection& iclIn, FramesCollection& iclOut){
//    iclOut.clear();
//    for(int i=0; i<iclIn.size(); ++i){
//        Frame f = iclIn[i];
//        iclOut.add(Frame(f,*this));
//    }
//}


void Pallet::readData(DataInputStream& dis) throw(IOException){
    int size = dis.readInt16();
    colors.clear();
    for(int i=0; i<size; ++i){
        IntVector3D key(dis);
        IntVector3D value(dis);
        colors[color2Int(key)] = color2Int(value);
    }
}

File: src/Editors/Editors Lib/src/pallet/Pallet.java
package pallet;

import file.AFileResolver;
import game.*;
import image.ImageUtils;
import java.awt.Color;
import java.awt.image.BufferedImage;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import util.UBTriple;

/**
 *
 * @author Gil Costa
 */
public class Pallet extends AFileResolver{
    //--------------
    // -- FIELDS --
    //--------------
    public HashMap<Color, Color> colors;
    
    //--------------------
    // -- CONSTRUCTORS --
    //--------------------
    public Pallet(String name){
        this();
        super.setFileName(GameInvariants.PALLETS_DIR + name + GameInvariants.PALLET_EXTENSION);
    }
    
    public Pallet(DataInputStream dis) throws IOException {
        this();
        this.readData(dis);
    }
    public Pallet(){
        super(GameInvariants.PALLET_EXTENSION, GameInvariants.PALLET_VERSION);
        colors = new HashMap<Color, Color>();
    }
    

    //---------------
    // -- METHODS --
    //---------------
    
    
    
     public BufferedImage applyToImage(BufferedImage img){
        BufferedImage target = ImageUtils.copyImage(img);
        for(int x=0; x<img.getWidth(); x++)
            for(int y=0; y<img.getHeight(); y++){
                // ignore transparent pixels
                if ((img.getRGB(x, y)&0xff000000) == 0)
                    continue;
                Color c = new Color(img.getRGB(x, y));
                if (c.getAlpha() == 0) continue;
                c = colors.get(c);
                if (c!=null){
                    int newRGB = c.getRGB();
                    target.setRGB(x, y, newRGB);
                }
            }
        return target;
    }
    
    
    
    
    
    
    
    protected void writeV1(DataOutputStream dos) throws IOException {
        
        dos.writeShort(colors.size());
        for(Map.Entry<Color, Color> entry:colors.entrySet()){
            Color c = entry.getKey();
            UBTriple key, value;
            key = new UBTriple(c.getRed(),c.getGreen(),c.getBlue());
            c = entry.getValue();
            value = new UBTriple(c.getRed(),c.getGreen(),c.getBlue());
            
            key.writeData(dos);
            value.writeData(dos);
        }
    }

    protected void readV1(DataInputStream dis) throws IOException {
        // v1.1
//        int size = dis.read();
//        char[] txt= new char[size];
//         for(int i=0; i<size; i++){
//            //txt[i] = dis.readChar();  // v1.0
//            txt[i] = (char)dis.readByte();  // v1.1
//         }
        
            
        int size = dis.readShort();
        colors = new HashMap<Color, Color>(size);
        for(int i=0; i<size; i++){
            UBTriple key = new UBTriple(dis);
            UBTriple value = new UBTriple(dis);
            colors.put(new Color(key.getX(),key.getY(),key.getZ()), new Color(value.getX(),value.getY(),value.getZ()));
        }
    }
    
    
    
    @Override
    public void writeData(DataOutputStream dos) throws IOException {
        switch(versionID){
            case 1: writeV1(dos); break;
            default: writeV1(dos);
        }
    }
     
    @Override
    public void readData(DataInputStream dis) throws IOException {
        switch(versionID){
            case 1: readV1(dis); break;
            default: readV1(dis);
        }
    }
    
    
}
